src/org/usfirst/frc4904/robot/commands/AutoShifter.java

```diff
@@ -0,0 +1,72 @@
+package org.usfirst.frc4904.robot.commands;
+
+
+import java.util.LinkedList;
+import org.usfirst.frc4904.robot.RobotMap;
+import org.usfirst.frc4904.standard.custom.sensors.CustomEncoder;
+import org.usfirst.frc4904.standard.subsystems.chassis.AutoSolenoidShifters;
+import org.usfirst.frc4904.standard.subsystems.chassis.SolenoidShifters;
+import edu.wpi.first.wpilibj.command.Command;
+
+public class AutoShifter extends Command {
+	protected final CustomEncoder leftEncoder;
+	protected final CustomEncoder rightEncoder;
+	protected final AutoSolenoidShifters shifter;
+	protected final LinkedList<Double> speeds = new LinkedList<Double>();
+	public static final double RAPID_ACCELERATION = 8;// WIP
+	public static final double RAPID_DECELERATION = 2;// WIP
+	public static final double RATE_DIFF = 100; // encoder ticks--WIP
+	public static final double SUPER_SLOW_RATE = 2; // ft per sec
+	public static final double MEDIUM_RATE = AutoShifter.SUPER_SLOW_RATE * 2;
+	public static final double FAST_RATE = AutoShifter.SUPER_SLOW_RATE * 4;
+	public static final double SLOW_THROTTLE = 0.4;
+	public static final double MEDIUM_THROTTLE = AutoShifter.SLOW_THROTTLE * 1.875;
+	public static final double FAST_THROTTLE = AutoShifter.SLOW_THROTTLE * 2;
+	public static final double SPEED_LIST_MAX_LENGTH = 5;
+	public static final double LAST_MANUAL_SHIFT_TIME = 5000; // TODO
+	public static final double LAST_AUTO_SHIFT_TIME = 500;// TODO
+
+	public AutoShifter(AutoSolenoidShifters shifter, CustomEncoder leftEncoder, CustomEncoder rightEncoder) {
+		this.shifter = shifter;
+		this.leftEncoder = leftEncoder;
+		this.rightEncoder = rightEncoder;
+	}
+
+	public AutoShifter() {
+		shifter = RobotMap.Component.shifter;
+		leftEncoder = RobotMap.Component.leftWheelEncoder;
+		rightEncoder = RobotMap.Component.rightWheelEncoder;
+	}
+
+	@Override
+	protected void execute() {
+		double speed = (leftEncoder.getRate() + rightEncoder.getRate()) / 2;
+		speeds.add(speed);
+		while (speeds.size() > AutoShifter.SPEED_LIST_MAX_LENGTH) {
+			speeds.removeFirst();
+		}
+		double speedDiffAvg = (speeds.getLast() - speeds.getFirst()) / (speeds.size() - 1);// acceleration
+		if (leftEncoder.getRate() - rightEncoder.getRate() < AutoShifter.RATE_DIFF
+			&& shifter.timeSinceLastManualShift() > AutoShifter.LAST_MANUAL_SHIFT_TIME
+			&& shifter.timeSinceLastAutoShift() > AutoShifter.LAST_AUTO_SHIFT_TIME) {// only if driving straight and we haven't manually shifted in 5 seconds or autoshifted in 0.5 seconds
+			double throttle = 0.0;
+			for (double motorSpeed : RobotMap.Component.chassis.getMotorSpeeds()) {
+				throttle += motorSpeed;
+			}
+			throttle /= RobotMap.Component.chassis.getMotorSpeeds().length;
+			if (Math.abs(speed) > AutoShifter.MEDIUM_RATE && Math.abs(speedDiffAvg) > AutoShifter.RAPID_ACCELERATION
```

Comment:
This condition will also trigger if the robot is skidding into another robot and you press the throttle up high and the wheels are accelerating. If the robot is hitting something, it should shift into low gear. One way around this is to also use acceleration from the navx. (Which I think you should add for sure.) I think that this code should be tested extensively on pinky when we get the shifters working again. Your units for throttle also are incorrect, I believe. I do not think that it is in units of feet per second. @HowardStark could you please confirm? Also meters per second?